From e9d8be82eac88c2585550cc2b860ecc026c032c2 Mon Sep 17 00:00:00 2001
From: Kevin Klues <klueska@gmail.com>
Date: Thu, 9 Feb 2017 19:39:42 -0800
Subject: [PATCH] Updated mesos containerizer to ignore GPU isolator creation
 failure.

This cherry-pick will be maintained in DC/OS to avoid the problem of
requiring NVML to be installed on *every* node in the cluster, even if
GPUs aren't available on them.

Without this patch, we would need some sort of per-agent configuration
to only enable the GPU isolator on agents that had NVML intstalled on
them, and that is currently difficult (or undesirable) to do in DC/OS
today.

This is a decent compromise.

Cherry-picked and modified from
`a9f5ddb8614ef057537994f389eb7e19f0c85652` due to changes introduced
by https://reviews.apache.org/r/62472/
---
 src/slave/containerizer/mesos/containerizer.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/slave/containerizer/mesos/containerizer.cpp b/src/slave/containerizer/mesos/containerizer.cpp
index c38bfacef..ec08818c4 100644
--- a/src/slave/containerizer/mesos/containerizer.cpp
+++ b/src/slave/containerizer/mesos/containerizer.cpp
@@ -443,8 +443,10 @@ Try<MesosContainerizer*> MesosContainerizer::create(
     {"gpu/nvidia",
       [&nvidia] (const Flags& flags) -> Try<Isolator*> {
         if (!nvml::isAvailable()) {
-          return Error("Cannot create the Nvidia GPU isolator:"
-                       " NVML is not available");
+          LOG(ERROR) << "Cannot create the Nvidia GPU isolator:"
+                        " NVML is not available";
+
+          return nullptr;
         }
 
         CHECK_SOME(nvidia)
@@ -504,7 +506,9 @@ Try<MesosContainerizer*> MesosContainerizer::create(
                    isolator.error());
     }
 
-    isolators.push_back(Owned<Isolator>(isolator.get()));
+    if (isolator.get() != nullptr) {
+      isolators.push_back(Owned<Isolator>(isolator.get()));
+    }
   }
 
   // Next, apply any custom isolators in the order given by the flags.
-- 
2.16.3

